### PR TITLE
Move Function/Module Lookup out of CaptureData

### PR DESCRIPTION
**Pull request by florian-kuebler**
This change allows function and module lookup from
places, where we don't have a capture available.
To do so, it moves the lookup functions from `CaptureData`
to a new set of free functions.

Further, this moves `ModuleManager` out of `CaptureData`
and store it explicitly.

The process and module information, as well as if needed
the capture information, now need to be passed explicitly
to the new lookup functions.

An immediate side effect of this change, is that
`CaptureData` becomes a pure data class again.

Test: Run Tests
Bug: http://b/216075140

### CODE REVIEW

File: src/ClientModel/SamplingDataPostProcessor.cpp [hunk 1]
```diff
@@ -125,13 +132,16 @@ class SamplingDataPostProcessor {
 
 PostProcessedSamplingData CreatePostProcessedSamplingData(const CallstackData& callstack_data,
                                                           const CaptureData& capture_data,
+                                                          const ModuleManager* module_manager,
```
Comment (dpallotti) [inline src/ClientModel/SamplingDataPostProcessor.cpp:135]: Why is `module_manager` passed by pointer while `callstack_data` and `capture_data` by reference? Can `module_manager` be `nullptr`? I thought we always have a `ModuleManager`. This applies also to `SamplingDataPostProcessor` itself.

File: src/ClientModel/SamplingDataPostProcessorTest.cpp [hunk 2]
```diff
@@ -594,27 +594,28 @@ class SamplingDataPostProcessorTest : public ::testing::Test {
...
-                                         100.0f, 0, 0.0f, kFunction3Instruction1AbsoluteAddress)),
-                                     SampledFunctionEq(MakeSampledFunction(
-                                         CaptureData::kUnknownFunctionOrModuleName,
-                                         CaptureData::kUnknownFunctionOrModuleName, 1, 20.0f, 1,
-                                         20.0f, 0, 0.0f, kFunction3Instruction2AbsoluteAddress)),
-                                     SampledFunctionEq(MakeSampledFunction(
-                                         CaptureData::kUnknownFunctionOrModuleName,
-                                         CaptureData::kUnknownFunctionOrModuleName, 1, 20.0f, 1,
-                                         20.0f, 0, 0.0f, kFunction4Instruction1AbsoluteAddress))));
+    EXPECT_THAT(
```
Comment (dpallotti) [inline src/ClientModel/SamplingDataPostProcessorTest.cpp:597]: clang-format?

File: src/ClientData/ModuleAndFunctionLookUp.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+#include <ObjectUtils/Address.h>
+
+using orbit_client_protos::FunctionInfo;
+using orbit_client_protos::LinuxAddressInfo;
+
+namespace orbit_client_data {
+namespace {
+[[nodiscard]] std::optional<uint64_t>
+FindFunctionAbsoluteAddressByInstructionAbsoluteAddressUsingModulesInMemory(
+    const ProcessData* process, const ModuleManager* module_manager, uint64_t absolute_address) {
```
Comment (dpallotti) [inline src/ClientData/ModuleAndFunctionLookUp.cpp:18]: Many of these functions take pointers to const, but assume the pointers are not `nullptr`. I think we should pass const references instead.

File: src/OrbitGl/App.h [hunk 4]
```diff
@@ -130,6 +130,16 @@ class OrbitApp final : public DataViewFactory,
     return *capture_data_;
   }
 
+  [[nodiscard]] const orbit_client_data::ModuleManager* GetModuleManager() const override {
```
Comment (dpallotti) [inline src/OrbitGl/App.h:133]: Why not return a const ref? `GetCaptureData` returns a const reference for example.

**Review by dpallotti — COMMENTED** [review]
The concept of the change looks great.

I'm fairly confused about the use of pointer to const instead of const references in many places (in particular when the pointer doesn't need to be stored in an object). These are often places where `CaptureData` or other classes were already const references, so the inconsistency stands out. Pointer to const over const reference tells me that we want to allow `nullptr` to be passed, but that doesn't seem to be the case.

**Review by beckerhe — APPROVED** [review]
That's an awesome change. I also wanted to move ModuleManager out of CaptureData for a while now. Thank you for doing this.

Davide raised some interesting questions. On top of that I only added some minor things. So I'm going to approve this.

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the review. I addressed all the comments.

File: src/OrbitGl/App.h (hunk 4 above)
Comment (florian-kuebler) [reply src/OrbitGl/App.h:133]: As discussed offline. It is stored in the tracks, so we keep it as a pointer.

File: src/DataViews/SamplingReportDataViewTest.cpp [hunk 5]
```diff
@@ -138,12 +144,16 @@ std::string GetExpectedDisplayAddressByIndex(size_t index) {
...
+std::string GetExpectedDisplayFunctionNameByIndex(size_t index, const ModuleManager* module_manager,
+                                                  CaptureData& capture_data) {
+  return orbit_client_data::GetFunctionNameByAddress(*module_manager, capture_data,
+                                                     kSampledAbsoluteAddresses[index]);
 }
 
-std::string GetExpectedDisplayModuleNameByIndex(size_t index, CaptureData& capture_data) {
-  std::string module_path = capture_data.GetModulePathByAddress(kSampledAbsoluteAddresses[index]);
+std::string GetExpectedDisplayModuleNameByIndex(size_t index, const ModuleManager* module_manager,
+                                                CaptureData& capture_data) {
```
Comment (dpallotti) [inline src/DataViews/SamplingReportDataViewTest.cpp:154]: Pass a const reference? Also can `capture_data` be a const ref as well (I know it was already like that)?

File: src/OrbitGl/CallTreeView.cpp [hunk 6]
```diff
@@ -118,14 +120,19 @@ uint64_t CallTreeNode::GetExclusiveSampleCount() const {
 static void AddCallstackToTopDownThread(CallTreeThread* thread_node,
                                         const CallstackInfo& resolved_callstack,
                                         uint64_t callstack_sample_count,
+                                        const ModuleManager* module_manager,
```
Comment (dpallotti) [inline src/OrbitGl/CallTreeView.cpp:123]: Can this be a const reference? Same over the entire file.

File: src/OrbitGl/CallTreeView.h [hunk 7]
```diff
@@ -143,11 +143,13 @@ class CallTreeView : public CallTreeNode {
  public:
   [[nodiscard]] static std::unique_ptr<CallTreeView> CreateTopDownViewFromPostProcessedSamplingData(
       const orbit_client_data::PostProcessedSamplingData& post_processed_sampling_data,
+      const orbit_client_data::ModuleManager* module_manager,
       const orbit_client_data::CaptureData& capture_data);
 
   [[nodiscard]] static std::unique_ptr<CallTreeView>
   CreateBottomUpViewFromPostProcessedSamplingData(
       const orbit_client_data::PostProcessedSamplingData& post_processed_sampling_data,
+      const orbit_client_data::ModuleManager* module_manager,
```
Comment (dpallotti) [inline src/OrbitGl/CallTreeView.h:152]: Yeah, mentioned in the other file, but here it's clearer that there inconsisteny in how the parameters are passed.

File: src/OrbitQt/CallTreeViewItemModelTest.cpp [hunk 8]
```diff
@@ -74,12 +74,13 @@ TEST(CallTreeViewItemModel, AbstractItemModelTesterFilledModel) {
...
+  orbit_client_data::ModuleManager module_manager{};
   orbit_client_data::PostProcessedSamplingData sampling_data =
       orbit_client_model::CreatePostProcessedSamplingData(capture_data->GetCallstackData(),
-                                                          *capture_data);
+                                                          *capture_data, module_manager);
 
-  auto call_tree_view =
-      CallTreeView::CreateTopDownViewFromPostProcessedSamplingData(sampling_data, *capture_data);
+  auto call_tree_view = CallTreeView::CreateTopDownViewFromPostProcessedSamplingData(
+      sampling_data, nullptr, *capture_data);
```
Comment (dpallotti) [inline src/OrbitQt/CallTreeViewItemModelTest.cpp:83]: Why not pass the `module_manager`?

File: src/OrbitQt/CallTreeViewItemModelTest.cpp [hunk 9]
```diff
@@ -92,11 +93,12 @@ TEST(CallTreeViewItemModel, SummaryItem) {
...
+    orbit_client_data::ModuleManager module_manager{};
     orbit_client_data::PostProcessedSamplingData sampling_data =
-        orbit_client_model::CreatePostProcessedSamplingData(capture_data->GetCallstackData(),
-                                                            *capture_data, generate_summary);
-    auto call_tree_view =
-        CallTreeView::CreateTopDownViewFromPostProcessedSamplingData(sampling_data, *capture_data);
+        orbit_client_model::CreatePostProcessedSamplingData(
+            capture_data->GetCallstackData(), *capture_data, module_manager, generate_summary);
+    auto call_tree_view = CallTreeView::CreateTopDownViewFromPostProcessedSamplingData(
+        sampling_data, nullptr, *capture_data);
```
Comment (dpallotti) [inline src/OrbitQt/CallTreeViewItemModelTest.cpp:101]: Same here.

**Review by dpallotti — COMMENTED** [review]
A couple of extra comments, more or less like the previous pass.

File: src/ClientData/ModuleAndFunctionLookup.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                                                             const std::string& build_id,
+                                                             uint64_t offset) {
+  const ModuleData* module_data = module_manager.GetModuleByPathAndBuildId(module_path, build_id);
+  if (module_data == nullptr) {
+    return nullptr;
+  }
+
+  uint64_t address = module_data->load_bias() + offset;
+
+  return module_data->FindFunctionByElfAddress(address, true);
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/ClientData/ModuleAndFunctionLookup.cpp]: Done. Thanks.

File: src/DataViews/SamplingReportDataViewTest.cpp (hunk 5 above)
Comment (florian-kuebler) [reply src/DataViews/SamplingReportDataViewTest.cpp]: Sure, good point. Missed that in the previous pass. Also made capture_data const ref.

File: src/OrbitGl/CallTreeView.cpp (hunk 6 above)
Comment (florian-kuebler) [reply src/OrbitGl/CallTreeView.cpp]: Good catch. Done.

File: src/OrbitGl/CallTreeView.h (hunk 7 above)
Comment (florian-kuebler) [reply src/OrbitGl/CallTreeView.h]: Agree. Done!

File: src/OrbitQt/CallTreeViewItemModelTest.cpp (hunk 8 above)
Comment (florian-kuebler) [reply src/OrbitQt/CallTreeViewItemModelTest.cpp]: That was not on purpose. Fixed it.

File: src/OrbitQt/CallTreeViewItemModelTest.cpp (hunk 9 above)
Comment (florian-kuebler) [reply src/OrbitQt/CallTreeViewItemModelTest.cpp]: Done.

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the second review. I (hopefully) addressed the comments.

**Review by dpallotti — APPROVED** [review]

**Pull request merged**